libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m

```diff
@@ -549,679 +452,290 @@ - (void)migrateUserDefaults {
-    return [userDefaults stringForKey:kUserDefaultsLastUserCommunityIdKey];
-}
+    BOOL userChanged = NO;
+    [accountsLock lock];
 
-- (void)setActiveCommunityId:(NSString *)activeCommunityId {
-    NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
-    if (activeCommunityId == nil) {
-        [userDefaults removeObjectForKey:kUserDefaultsLastUserCommunityIdKey];
+    if (nil==user) { //clear current user if  nil
+        _currentUser = nil;
+        [self setCurrentUserIdentity:nil];
+        userChanged = YES;
     } else {
-        [userDefaults setObject:activeCommunityId forKey:kUserDefaultsLastUserCommunityIdKey];
+        //check if this is valid managed user
+        SFUserAccount *userAccount = [self userAccountForUserIdentity:user.accountIdentity];
+        if (nil!=userAccount) {
+          [self willChangeValueForKey:@"currentUser"];
+               _currentUser = user;
+            [self setCurrentUserIdentity:user.accountIdentity];
+          userChanged = YES;
+          [self didChangeValueForKey:@"currentUser"];
+        } else {
+          [self log:SFLogLevelError format:@"Cannot set the currentUser as %@. Add the account to the SFAccountManager before making this call.",[user userName]];
+        }
     }
-    [userDefaults synchronize];
+
+    [accountsLock unlock];
+
+    if (userChanged)
+        [self userChanged:SFUserAccountChangeUnknown];
 }
 
-- (void)setActiveUser:(SFUserAccount *)user {
-    if (nil == user) {
-        self.activeCommunityId = nil;
-        self.activeUserIdentity = nil;
+-(SFUserAccountIdentity *) currentUserIdentity {
+    SFUserAccountIdentity *accountIdentity = nil;
+    [accountsLock lock];
+    if (_currentUser == nil) {
+        NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
+        accountIdentity = [userDefaults objectForKey:kUserDefaultsLastUserIdentityKey];
     } else {
-        SFUserAccountIdentity *userIdentity = user.accountIdentity;
-        self.activeUserIdentity = userIdentity;
-        self.activeCommunityId = user.communityId;
+        accountIdentity = _currentUser.accountIdentity;
     }
+    [accountsLock unlock];
+    return accountIdentity;
 }
 
-- (void)replaceOldUser:(SFUserAccountIdentity *)oldUserIdentity withUser:(SFUserAccount *)newUser {
-    [self.userAccountMap removeObjectForKey:oldUserIdentity];
-    SFUserAccountIdentity *newIdentity = newUser.accountIdentity;
-    (self.userAccountMap)[newIdentity] = newUser;
-    
-    SFUserAccountIdentity *defUserIdentity = self.activeUserIdentity;
-    if (!defUserIdentity || [defUserIdentity isEqual:oldUserIdentity]) {
-        [self setActiveUser:newUser];
+- (void)setCurrentUserIdentity:(SFUserAccountIdentity*)userAccountIdentity {
+    NSUserDefaults *standardDefaults = [NSUserDefaults msdkUserDefaults];
+    [accountsLock lock];
+    if (userAccountIdentity!=nil) {  //clear current user if userAccountIdentity is nil
```

Comment:
Minor: I'd probably put the comment on the `else` block line, since it applies there.